### DISCUSSION TITLE
How to change the order of joints in pinocchio

### DISCUSSION THREAD

**Discussion by nanyangnuan** [stack-of-tasks/pinocchio#2209] [Q&A]
Hello! When I use 'buildmodel' to build unitree_A1, the order of joints is 'FL,FR,RL,RR' in pinocchio's model. But the order of input is 'FR,FL,RR,RL'. So, how to change the order of joints if the order of input can't change? Thanks in advance.

**Comment by jcarpent** [comment]
Why do you want to change the joint ordering?

**Reply by nanyangnuan** [reply]
when I use unitree_guide, the order of input just like position q is 'FR,FL,RR,RL'. I use pinocchio to buildmodel and want to forwardKinematics(model,data,_q), the order of _q is 'FL,FR,RL,RR' , I need change q to _q.  other parameter like v,a,tau and so on need to change order. I think this work is pointless.

**Reply by jcarpent** [reply]
Currently, we do not have such tools to reorder the joints as you mention. It would be quite nice to have indeed.
In the current time, I would suggest just doing the overwriting manually.

**Reply by cmastalli** [reply]
@nanyangnuan -- your request is very user-specific and not well thought out. Pinocchio follows an alphanumeric order as in other well-developed cases. You can easily reorder joints to accommodate your requirements.

**Reply by york-fu** [reply]
Pinocchio determines the joint order through depth-first traversal but arranges sibling chains alphabetically at branch points, which affects the order of q. When loading a URDF file, the URDF file is structured from top to bottom, naturally defining one chain first and then another, clearly specifying the order of sibling chains at branch points. Why not follow this order for arrangement?

**Reply by MegMll** [reply]
Hello, 

In pinocchio, to parse a urdf file, we use liburdfdom, which is the one ordering the joints alphabetically. 
So unfortunately as of now, if you want to re-order your joints, you can either rename your joints, or you can create smaller urdf files, create smaller models, and use the function `appendModel` to append the chains and branch as you want.
